local-proxy/src/index.ts

```diff
@@ -29,33 +32,48 @@ export async function startProxy(rootDir: string, localToken: string) {
       SHIFT_DEV_SERVER_BASE_REQUIRE_PATH: path.resolve(path.join(rootDir, 'backend')),
       SHIFT_DEV_SERVER_LOCAL_TOKEN: localToken,
     },
+    // Workaround for tests:
+    // server.js uses babel/dir which has a console.log
+    // AVA transaltes console.log to stderr
+    // Rush detects stderr as warnings and returns an exit code
+    stdout: process.env.NODE_ENV !== 'test',
   });
 
-  let handle: any;
-
   nodemon.on('quit', () => {
     process.exit();
-  }).on('start', (child) => {
-    handle = child;
+  }).on('start', (_child) => {
+    if (process.env.NODE_ENV !== 'test') {
+      log('Local dev server started');
+    }
   }).on('message', (message) => {
-    if (message === 'ready' && handle !== undefined) {
-      handle.send('server', server);
+    if (message.type === 'ready') {
+      port = message.port;
+      readyEmitter.emit('ready');
     }
   }).on('restart', (files) => {
-    console.log('Local dev server restarted due to changes in: ', files);
+    port = undefined;
+    log('Local dev server restarted due to changes in: ', files);
   });
-
-  return port;
 }
 
 export function setupProxy(sourceDir: string) {
   const rootDir = path.resolve(sourceDir, '..');
   const localToken = nanoid();
-  return async (app: Application) => {
-    const port = await startProxy(rootDir, localToken);
-    app.use(proxy('/invoke', {
-      target: `http://localhost:${port}/`,
-      headers: { 'x-shift-dev-server-local-token': localToken },
-    }));
+  const httpProxy = new proxy();
+  httpProxy.on('error', (err: any) => console.error(err.stack));
+  return (app: Application) => {
+    startProxy(rootDir, localToken);
+    app.use(async (req, res, next) => {
+      if (req.url !== '/invoke') {
+        return next();
+      }
+      if (!port) {
+        await new Promise((resolve) => readyEmitter.once('ready', resolve));
```

Comment:
Given that it's `once`, it might be easier to read if you defined the promise once and just `await`ed that same promise each time here.